File: mobile/lib/models/work_sample.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'work_sample.freezed.dart';
part 'work_sample.g.dart';

/// Sync status for work samples
enum SyncStatus {
  @JsonValue('local')
  local, // Stored locally only
  @JsonValue('syncing')
  syncing, // Currently uploading
  @JsonValue('synced')
  synced, // Uploaded to R2
  @JsonValue('error')
  error, // Upload failed
}

/// A work sample (photo, document, etc.) attached to a log entry.
/// 
/// For offline-first mobile:
/// 1. Files are stored locally first in the backup directory
/// 2. If sync is enabled, files are uploaded to R2 when on good network
/// 3. [localPath] stores the local file path
/// 4. [storageKey] stores the R2 object key (after sync)
@freezed
class WorkSample with _$WorkSample {
  const factory WorkSample({
    required String id,
    required String familyId,
    required String logEntryId,
    String? groupId, // Links to all logs in a multi-student group
    required String studentId,
    required String fileName,
    String? localPath, // Local file path (for offline storage)
    String? storageKey, // R2 object key (after sync)
    required String contentType,
    required int sizeBytes,
    required String uploadedBy,
    String? description,
    @Default(SyncStatus.local) SyncStatus syncStatus,
    String? downloadUrl, // Pre-signed URL for viewing
    String? expiresAt, // URL expiration
    required DateTime createdAt,
  }) = _WorkSample;

  factory WorkSample.fromJson(Map<String, dynamic> json) =>
      _$WorkSampleFromJson(json);
}

/// Premium features for a family
@freezed
class PremiumFeatures with _$PremiumFeatures {
  const factory PremiumFeatures({
    @Default(false) bool syncEnabled,
    @Default(false) bool uploadsEnabled,
    @Default(0) int storageUsedBytes,
    @Default(104857600) int storageLimitBytes, // 100MB default
    @Default('free') String subscriptionTier,
    DateTime? subscriptionEnd,
  }) = _PremiumFeatures;

  factory PremiumFeatures.fromJson(Map<String, dynamic> json) =>
      _$PremiumFeaturesFromJson(json);
}

/// Storage usage information
@freezed
class StorageUsage with _$StorageUsage {
  const factory StorageUsage({
    required int usedBytes,
    required int limitBytes,
    required double usedPercent,
    required int fileCount,
    required bool uploadsEnabled,
    required double usedMB,
    required double limitMB,
    required int remainingBytes,
  }) = _StorageUsage;

  factory StorageUsage.fromJson(Map<String, dynamic> json) =>
      _$StorageUsageFromJson(json);
}

/// Response when requesting an upload URL
@freezed
class UploadUrlResponse with _$UploadUrlResponse {
  const factory UploadUrlResponse({
    required String uploadUrl,
    required String workSampleId,
    required String storageKey,
    required String expiresAt,
  }) = _UploadUrlResponse;

  factory UploadUrlResponse.fromJson(Map<String, dynamic> json) =>
      _$UploadUrlResponseFromJson(json);
}

/// Allowed content types for work sample uploads
const allowedContentTypes = {
  'image/jpeg',
  'image/png',
  'image/gif',
  'image/webp',
  'image/heic',
  'image/heif',
  'application/pdf',
  'video/mp4',
  'video/quicktime',
};

/// Maximum file size (10MB)
const maxFileSizeBytes = 10 * 1024 * 1024;

/// Check if a content type is allowed
bool isAllowedContentType(String contentType) {
  return allowedContentTypes.contains(contentType);
}

File: backend/models/work_sample.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// WorkSample represents a file attached to a log entry (photo, document, etc.)
// For multi-student logs (same groupId), work samples are shared across all logs in the group.
type WorkSample struct {
	ID          primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	FamilyID    primitive.ObjectID `bson:"familyId" json:"familyId"`
	LogEntryID  primitive.ObjectID `bson:"logEntryId" json:"logEntryId"`               // Primary log entry
	GroupID     *string            `bson:"groupId,omitempty" json:"groupId,omitempty"` // Links to all logs in group
	StudentID   primitive.ObjectID `bson:"studentId" json:"studentId"`                 // Primary student (uploader's context)
	FileName    string             `bson:"fileName" json:"fileName"`                   // Original filename
	StorageKey  string             `bson:"storageKey" json:"storageKey"`               // R2 object key (or local path for mobile)
	ContentType string             `bson:"contentType" json:"contentType"`             // MIME type
	SizeBytes   int64              `bson:"sizeBytes" json:"sizeBytes"`
	UploadedBy  primitive.ObjectID `bson:"uploadedBy" json:"uploadedBy"` // User who uploaded
	Description string             `bson:"description,omitempty" json:"description,omitempty"`
	SyncStatus  string             `bson:"syncStatus,omitempty" json:"syncStatus,omitempty"` // "local", "syncing", "synced"
	CreatedAt   time.Time          `bson:"createdAt" json:"createdAt"`
}

// WorkSampleWithURL includes a pre-signed download URL for client access
type WorkSampleWithURL struct {
	WorkSample
	DownloadURL string `json:"downloadUrl"`
	ExpiresAt   string `json:"expiresAt"`
}

// AllowedContentTypes for work sample uploads
var AllowedContentTypes = map[string]bool{
	// Images
	"image/jpeg": true,
	"image/png":  true,
	"image/gif":  true,
	"image/webp": true,
	"image/heic": true,
	"image/heif": true,

	// Documents
	"application/pdf": true,

	// Videos (optional, larger files)
	"video/mp4":       true,
	"video/quicktime": true,
}

// MaxFileSizeBytes is the default max file size (10MB)
const MaxFileSizeBytes = 10 * 1024 * 1024

// DefaultStorageLimitBytes is the default storage limit per family (100MB for free tier)
const DefaultStorageLimitBytes = 100 * 1024 * 1024

// PremiumStorageLimitBytes is the storage limit for premium tier (5GB)
const PremiumStorageLimitBytes = 5 * 1024 * 1024 * 1024

// IsAllowedContentType checks if a content type is allowed for upload
func IsAllowedContentType(contentType string) bool {
	return AllowedContentTypes[contentType]
}

// GetStorageKey generates the R2 storage key for a work sample
func GetStorageKey(familyID, logEntryID primitive.ObjectID, uuid, fileName string) string {
	return "families/" + familyID.Hex() + "/work-samples/" + logEntryID.Hex() + "/" + uuid + "-" + fileName
}
